### DISCUSSION TITLE
How to read from an endpoint pipe?

### DISCUSSION THREAD

**Discussion by marcos-diaz** [hathach/tinyusb#1088] [Q&A] [answered]
Hi, using TinyUSB (as device) in a Raspberry Pico, I created a custom app driver to deal with an interface with 2 endpoints (IN and OUT). I'm opening the endpoints as:
```
my_driver_open(uint8_t rhport, tusb_desc_interface_t const *itf_desc, uint16_t max_len) {
    ...
    usbd_edpt_open(rhport, ep_in);
    usbd_edpt_open(rhport, ep_out);
``` 

and successfully sending data to the IN endpoint as:
```
report_in(uint8_t* report) {
    ...
    usbd_edpt_claim(0, addr);
    usbd_edpt_xfer(0, addr, report, REPORT_SIZE);
    usbd_edpt_release(0, addr);
```

Now, how should I implement the OUT callback to receive data? I was hoping that the function assigned to `.xfer_cb` was called but it is not.

I also checked that the endpoint is correctly detected as `tu_edpt_dir(addr) == TUSB_DIR_OUT`, so it looks correct.

Could you please give me some hints, or point me to a similar piece of code on how to receive data from an endpoint?

Thanks!

**Comment by HiFiPhile** [comment, marked as answer]
You can try vendor class (`class/vendor_device`), it handles a pair of bulk endpoints. If you want is raw bulk transfer you don't need a custom driver.

**Reply by jmark1m** [reply]
I'm trying to do the exact same thing as marcos-diaz and bulk transfer (write / read) as a vendor host.

Can you provide some help on the use of ( class/vendor_host )?

To use it I set the CFG_TUH_VENDOR define.  But then upon compiling it, it complains about "pipe_handle_t" was not declared.

But "pipe_handle_t" is not defined anywhere in the tinyusb project.
